File: sdk/typescript/src/embed.ts
import { spawn, ChildProcess } from "node:child_process";
import { existsSync, statSync, accessSync, constants as fsConstants } from "node:fs";
import { dirname, resolve as pathResolve } from "node:path";
import { createInterface } from "node:readline";
import { BinaryNotFoundError } from "./index.js";

/** 4-step binary discovery, mirroring python/beava/_embed.py. */
export function discoverBinary(): string {
  // Step 1: BEAVA_BINARY env var — explicit override; MUST be valid if set.
  const envVal = process.env.BEAVA_BINARY;
  if (envVal !== undefined && envVal !== "") {
    if (existsSync(envVal) && statSync(envVal).isFile()) {
      try {
        accessSync(envVal, fsConstants.X_OK);
        return envVal;
      } catch {
        /* fall through to error */
      }
    }
    throw new BinaryNotFoundError(
      `BEAVA_BINARY=${JSON.stringify(envVal)} is set but the path is not an executable file. Unset BEAVA_BINARY or fix the path.`,
      [envVal],
    );
  }

  // Step 2: beava on PATH.
  const pathDirs = (process.env.PATH ?? "").split(":").filter(Boolean);
  for (const dir of pathDirs) {
    const candidate = `${dir}/beava`;
    if (existsSync(candidate) && statSync(candidate).isFile()) {
      try {
        accessSync(candidate, fsConstants.X_OK);
        return candidate;
      } catch {
        continue;
      }
    }
  }

  // Step 3: Walk upward from CWD looking for target/debug/beava.
  const searched: string[] = [];
  let dir = process.cwd();
  while (true) {
    const candidate = pathResolve(dir, "target/debug/beava");
    searched.push(candidate);
    if (existsSync(candidate) && statSync(candidate).isFile()) {
      try {
        accessSync(candidate, fsConstants.X_OK);
        return candidate;
      } catch {
        /* not executable */
      }
    }
    const parent = dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }

  // Step 4: Not found.
  throw new BinaryNotFoundError(
    "beava binary not found. Install with one of:\n" +
      "  brew install beava\n" +
      "  pip install beava[server]\n" +
      "  docker pull beava/beava\n" +
      "Or set BEAVA_BINARY=/path/to/beava.",
    searched,
  );
}

export interface SpawnedServer {
  proc: ChildProcess;
  httpUrl: string;
  tcpUrl: string;
}

export interface SpawnOptions {
  startupTimeoutMs?: number;
  testMode?: boolean;
}

/** Spawn a local beava server on ephemeral ports and wait until both bind events appear. */
export async function spawnEmbeddedServer(opts: SpawnOptions = {}): Promise<SpawnedServer> {
  const timeoutMs = opts.startupTimeoutMs ?? 5000;
  const binary = discoverBinary();

  const env: NodeJS.ProcessEnv = {
    ...process.env,
    BEAVA_LISTEN_ADDR: "127.0.0.1:0",
    BEAVA_TCP_PORT: "0",
    BEAVA_DEV_ENDPOINTS: "1",
  };
  if (opts.testMode) {
    env.BEAVA_TEST_MODE = "1";
  }

  const proc = spawn(binary, ["--config", "/dev/null"], {
    env,
    stdio: ["ignore", "pipe", "ignore"],
  });

  let httpAddr: string | null = null;
  let tcpAddr: string | null = null;

  const promise = new Promise<SpawnedServer>((resolve, reject) => {
    const rl = createInterface({ input: proc.stdout! });
    const timer = setTimeout(() => {
      rl.close();
      try {
        proc.kill("SIGKILL");
      } catch {
        /* ignore */
      }
      reject(
        new Error(
          `embed-mode server did not bind within ${timeoutMs}ms (http=${httpAddr ?? "null"}, tcp=${tcpAddr ?? "null"})`,
        ),
      );
    }, timeoutMs);

    rl.on("line", (line) => {
      try {
        const rec = JSON.parse(line) as { kind?: string; addr?: string };
        if (rec.kind === "server.http_bound" && rec.addr) {
          httpAddr = rec.addr;
        } else if (rec.kind === "server.tcp_bound" && rec.addr) {
          tcpAddr = rec.addr;
        }
        if (httpAddr && tcpAddr) {
          clearTimeout(timer);
          resolve({
            proc,
            httpUrl: `http://${httpAddr}`,
            tcpUrl: `tcp://${tcpAddr}`,
          });
        }
      } catch {
        // non-JSON line (banner) — ignore during startup
      }
    });
    proc.on("exit", (code) => {
      if (!httpAddr || !tcpAddr) {
        clearTimeout(timer);
        reject(new Error(`embed beava exited (code=${code}) before binding`));
      }
    });
  });

  return promise;
}

/** Send SIGTERM, wait, SIGKILL on timeout. */
export async function teardownProcess(
  proc: ChildProcess,
  timeoutMs = 5000,
): Promise<void> {
  if (proc.exitCode !== null) return;
  return new Promise<void>((resolve) => {
    const timer = setTimeout(() => {
      try {
        proc.kill("SIGKILL");
      } catch {
        /* ignore */
      }
    }, timeoutMs);
    proc.once("exit", () => {
      clearTimeout(timer);
      resolve();
    });
    try {
      proc.kill("SIGTERM");
    } catch {
      clearTimeout(timer);
      resolve();
    }
  });
}

File: sdk/go/embed.go
package beava

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"sync"
	"syscall"
	"time"
)

// discoverBinary locates the beava binary using the 4-step order documented
// in python/beava/_embed.py.
func discoverBinary() (string, error) {
	searched := []string{}

	// Step 1: BEAVA_BINARY env var.
	if envVal := os.Getenv("BEAVA_BINARY"); envVal != "" {
		searched = append(searched, envVal)
		if info, err := os.Stat(envVal); err == nil && !info.IsDir() && info.Mode()&0o111 != 0 {
			return envVal, nil
		}
		return "", &BinaryNotFoundError{
			Searched: searched,
			Reason:   fmt.Sprintf("BEAVA_BINARY=%q is set but the path is not an executable file", envVal),
		}
	}

	// Step 2: beava on PATH.
	if p, err := exec.LookPath("beava"); err == nil {
		return p, nil
	}

	// Step 3: walk parents for target/debug/beava.
	cwd, _ := os.Getwd()
	dir := cwd
	for {
		candidate := filepath.Join(dir, "target", "debug", "beava")
		searched = append(searched, candidate)
		if info, err := os.Stat(candidate); err == nil && !info.IsDir() && info.Mode()&0o111 != 0 {
			return candidate, nil
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}

	return "", &BinaryNotFoundError{
		Searched: searched,
		Reason: "beava binary not found. Install with one of:\n" +
			"  brew install beava\n" +
			"  pip install beava[server]\n" +
			"  docker pull beava/beava\n" +
			"Or set BEAVA_BINARY=/path/to/beava.",
	}
}

// SpawnedServer holds the spawned embed-mode process plus discovered URLs.
type SpawnedServer struct {
	Cmd     *exec.Cmd
	HTTPURL string
	TCPURL  string

	stdoutPipe io.ReadCloser
	once       sync.Once
}

// SpawnOptions configure embed-mode startup.
type SpawnOptions struct {
	StartupTimeout time.Duration // default 5s
	TestMode       bool
}

// SpawnEmbeddedServer launches a local beava server on ephemeral ports and
// blocks until both server.http_bound and server.tcp_bound events arrive.
func SpawnEmbeddedServer(ctx context.Context, opts SpawnOptions) (*SpawnedServer, error) {
	timeout := opts.StartupTimeout
	if timeout == 0 {
		timeout = 5 * time.Second
	}
	binary, err := discoverBinary()
	if err != nil {
		return nil, err
	}

	cmd := exec.CommandContext(ctx, binary, "--config", "/dev/null")
	cmd.Env = append(os.Environ(),
		"BEAVA_LISTEN_ADDR=127.0.0.1:0",
		"BEAVA_TCP_PORT=0",
		"BEAVA_DEV_ENDPOINTS=1",
	)
	if opts.TestMode {
		cmd.Env = append(cmd.Env, "BEAVA_TEST_MODE=1")
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, fmt.Errorf("beava: stdout pipe: %w", err)
	}
	cmd.Stderr = nil

	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("beava: spawn: %w", err)
	}

	type result struct {
		http, tcp string
		err       error
	}
	resCh := make(chan result, 1)

	go func() {
		scanner := bufio.NewScanner(stdout)
		var httpAddr, tcpAddr string
		for scanner.Scan() {
			line := scanner.Bytes()
			var rec struct {
				Kind string `json:"kind"`
				Addr string `json:"addr"`
			}
			if err := json.Unmarshal(line, &rec); err != nil {
				continue
			}
			switch rec.Kind {
			case "server.http_bound":
				httpAddr = rec.Addr
			case "server.tcp_bound":
				tcpAddr = rec.Addr
			}
			if httpAddr != "" && tcpAddr != "" {
				resCh <- result{http: httpAddr, tcp: tcpAddr}
				return
			}
		}
		resCh <- result{err: fmt.Errorf("beava: stdout closed before bind events arrived")}
	}()

	timer := time.NewTimer(timeout)
	defer timer.Stop()
	select {
	case r := <-resCh:
		if r.err != nil {
			_ = cmd.Process.Kill()
			return nil, r.err
		}
		return &SpawnedServer{
			Cmd:        cmd,
			HTTPURL:    "http://" + r.http,
			TCPURL:     "tcp://" + r.tcp,
			stdoutPipe: stdout,
		}, nil
	case <-timer.C:
		_ = cmd.Process.Kill()
		return nil, fmt.Errorf("beava: embed-mode server did not bind within %s", timeout)
	case <-ctx.Done():
		_ = cmd.Process.Kill()
		return nil, ctx.Err()
	}
}

// Teardown sends SIGTERM, then SIGKILL after `timeout`.
func (s *SpawnedServer) Teardown(timeout time.Duration) error {
	if s == nil || s.Cmd == nil || s.Cmd.Process == nil {
		return nil
	}
	if timeout == 0 {
		timeout = 5 * time.Second
	}
	_ = s.Cmd.Process.Signal(syscall.SIGTERM)
	done := make(chan error, 1)
	go func() { done <- s.Cmd.Wait() }()
	select {
	case <-done:
		return nil
	case <-time.After(timeout):
		_ = s.Cmd.Process.Kill()
		<-done
		return nil
	}
}
